docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,198 @@
+
+For the first pass, we would likely limit the number of kbparallel runs.
+
+For the next pass, we would want to create a comprehensive generalized solution to submit,split and aggregate, with recipes or conveniences for common operations for creating sets, reports, or things of that nature.
+
+We would also want to do a user study on what we want from the UI and which functionality we want, as the UI may inform the design of the backend system.
+
+
+### Deprecate KBP and instead break out apps into 3 parts
+
+* Fan out (FO)
+* Process in parallel (PIP)
+* Fan in (FI)
+
+
+### Steps:
+1. User launches job as normal
+2. Possibly the job is marked as a FO job, Makes it easier for the UI to display the job correctly initially, Ideally would be marked in the spec, but this might be a lot of work Could potentially be marked in the catalog UI (e.g. along with the job requirements) 
+3. Job figures out what the PIP / sub jobs should be
+4. Job sends the following info to EE2
+* Its own job ID
+* The parameters for each of the sub jobs
+* The app of the FI job, e.g. kb_phylogenomics/build_microbial_speciestree_reduce
+* EE2 starts the subjobs and associates them with with FO job (Probably need retry handling for the subjobs to deal with transient errors)
+5. Whenever a subjob finishes, EE2 checks to see if all the subjobs are finished
+* If true, EE2 starts the FI job, providing the outputs of the subjobs as a list to the reduce job
+* When the FI job is finished, the job is done.
+* The various jobs can communicate by storing temporary data in the caching service or in the Blobstore. If the latter is used, the FI job should clean up the Blobstore nodes when its complete.
+* Could make a helper app for this?
+* What about workflow engines (WDL, Cromwell)? Are we reinventing the wheel here?
+* Can new EE2 endpoints speed up or reduce the complexity of any of these steps?
+
+### Notes about DAG in ee2 Endpoints
+```
+Your dag would need to have (at least) a first job followed by a SUBDAG EXTERNAL.
+Somewhere in the first job you'd generate a new dag workflow that
+defines the N clusters followed by the N+1 job, which runs in the
+subdag.
+
+As for DAGMan support in the Python bindings, we do this in the
+following two ways:
+
+1) There is a htcondor.Submit.from_dag() option which takes the name
+of a dag filename. You then submit the resulting object just like any
+regular job.
+2) We have a htcondor.dags library which can be used to
+programmatically construct a DAG workflow in computer memory, then
+write to a .dag file and submit using the function mentioned in 1)
+above.
+```
+
+Between these there are several different ways to do what you want.
+
+There's a useful example here that shows the general workflow in the
+bindings: https://htcondor.readthedocs.io/en/latest/apis/python-bindings/tutorials/DAG-Creation-And-Submission.html#Describing-the-DAG-using-htcondor.dags
+
+## Consequences
+
+* We will have to roll out fixes in multiple stages
```

Comment:
Why is this?

Reply:
Because Decision Outcome  has a solution with multiple stages of fixes.

Reply:
Oh, you're talking about the quick fix vs. long term fix thing. We don't *have* to do that though - we could just go straight for the long term fix.

Reply:
True. Do you want me to remove this line?

Reply:
I guess so, it confused me anyway, and it's not strictly true.

Reply:
Ok, then I will remove it

Reply:
:+1: